shared/utils/logic.ts

```diff
@@ -265,102 +312,137 @@ export const getVisibleFieldIds = (
-    condition.state === LogicConditionState.Either
-  ) {
-    // condition.value can be a string (is equals to), or an array (is either) (not strictly true either...)
-    const conditionValues = ([] as unknown[])
-      .concat(condition.value)
-      .map(String)
-      // TODO #4279: Revisit decision to trim in backend after React rollout is complete
-      .map((opt) => opt.trim())
-    currentValue = String(currentValue)
-    /*
-    Handling 'Others' for radiobutton
-
-    form_logics: [{ ... value : 'Others' }]
-
-    Client-side:
-    When an Others radiobutton is checked, the fieldValue is 'radioButtonOthers'
-
-    Server-side:
-    When an Others radiobutton is checked, and submitted with the required value,
-    the answer is: 'Others: value'
-    */
-
-    // TODO: An option that is named "Others: Something..." will also pass this test,
-    // even if the field has not been configured to set othersRadioButton=true
-    if (conditionValues.indexOf('Others') > -1) {
-      // TODO: This is used for angular's client. The server has no need for these magic values, as they are never seen by the server.
-      if (field.fieldType === 'radiobutton') {
-        conditionValues.push('radioButtonOthers')
-      } else if (field.fieldType === 'checkbox') {
-        conditionValues.push('checkboxOthers') // Checkbox currently doesn't have logic, but the 'Others' will work in the future if it in implemented
+  if (typeof currentValue !== 'string') {
+    // Handle the case where the value is a client radio field value.
+    currentValue = {
+      ...currentValue,
+      value: currentValue.value?.trim(),
+    }
+  }
+
+  // Evaluate condition
+  const trimmedCondition = trimConditionValue(condition)
+
+  switch (trimmedCondition.state) {
+    case LogicConditionState.Lte:
+      return Number(currentValue) <= Number(trimmedCondition.value)
+    case LogicConditionState.Gte:
+      return Number(currentValue) >= Number(trimmedCondition.value)
+    case LogicConditionState.Equal:
+    case LogicConditionState.Either: {
+      const conditionValues = Array.isArray(trimmedCondition.value)
+        ? trimmedCondition.value
+        : [trimmedCondition.value]
+
+      /**
+       * Field type   Radio       Decimal     Everything else
+       * Server-side  (1)         (2)         (3)
+       * Client-side  (4)         (5)         (6)
+       *
+       * (1) If conditionValues includes "Others", must check for radio answer
+       *     starting with "Others: ".
+       *     If not, continue the usual string-based check with (3).
+       * (4) If conditionValues includes "Others", check for the value being
+       *     special radio value and the othersInput subfield having a value.
+       *     If not, continue the usual string-based check with (3).
+       * (2) + (5) Use Number equality for decimals
+       * (3) + (6) Use String equality for everything else
+       */
+
+      // Special radio others check
+      if (
```

Comment:
I had quite a hard time matching 1-6 over to each logic.

Does it make sense to indicate which case each control statement is handling?

Reply:
thanks for pointing it out! Added here [86e5e93](https://github.com/opengovsg/FormSG/pull/7136/commits/86e5e937ff3ebece7ec249bcc8ebeab48f0fd162)